resnet/data.py

```diff
@@ -0,0 +1,219 @@
+# If using ToTensor() and the default collate, remove the scaling by 255
+IMAGENET_CHANNEL_MEAN = (0.485 * 255, 0.456 * 255, 0.406 * 255)
+IMAGENET_CHANNEL_STD = (0.229 * 255, 0.224 * 255, 0.225 * 255)
+
+class StreamingImageNet(Dataset, VisionDataset):
+    """ ImageNet streaming dataset based on PyTorch's VisionDataset.
+
+    Args:
+        remote (str): Remote directory (S3 or local filesystem) where dataset is stored.
+        local (str): Local filesystem directory where dataset is cached during operation.
+        split (str): The dataset split to use, either 'train' or 'val'.
+        shuffle (bool): Whether to iterate over the samples in randomized order.
+        transform (callable, optional): A function/transform that takes in an image and returns a transformed version.
+            Default: ``None``.
+        prefetch (int, optional): Target number of samples remaining to prefetch while iterating.
+            Default: ``100_000``.
+        retry (int, optional): Number of download re-attempts before giving up. Defaults to ``2``.
+        timeout (float, optional): Number of seconds to wait for a shard to download before raising
+            an exception. Default: 120 seconds.
+        batch_size (int, optional): The batch size that will be used on each device's DataLoader.
+            Default: ``None``.
+    """
+
+    def __init__(self,
+                 remote: str,
+                 local: str,
+                 split: str,
+                 shuffle: bool,
+                 transform: Optional[Callable] = None,
+                 prefetch: Optional[int] = 100_000,
+                 retry: int = 2,
+                 timeout: float = 120,
+                 batch_size: Optional[int] = None) -> None:
+
+        if split not in ['train', 'val']:
+            raise ValueError(
+                f"split='{split}', but must be one of ['train', 'val'].")
+
+        self.transform = transform
+
+        super().__init__(remote=remote,
+                         local=local,
+                         split=split,
+                         shuffle=shuffle,
+                         prefetch=prefetch,
+                         keep_zip=False,
+                         retry=retry,
+                         timeout=timeout,
+                         hash=None,
+                         batch_size=batch_size)
+
+    def __getitem__(self, idx: int) -> Any:
+        sample = super().__getitem__(idx)
+        image = sample['x']
+        if image.mode != "RGB":
+            image = image.convert('RGB')
+        if self.transform:
+            image = self.transform(image)
+        target = sample['y']
+        return image, target
+
+
+def build_streaming_imagenet_dataspec(
+    remote: str,
+    local: str,
+    batch_size: int,
+    is_train: bool = True,
+    drop_last: bool = True,
+    shuffle: bool = True,
+    resize_size: int = -1,
+    crop_size: int = 224,
+    **dataloader_kwargs,
+) -> DataSpec:
```

Comment:
can we deduplicate `build_streaming_imagenet_dataspec` and `build_imagenet_dataspec`? You've thought about it more but it seems like we could just add a `streaming: bool = False` argument and just change the dataset type near the bottom?

Would save us and the user some code + make clear that using streaming datasets is barely different from what you're used to.

Reply:
Either works, I thought it was nice to have a clear distinction between using the two, but to your point, it might be good to blur the lines

Reply:
Done!